### DISCUSSION TITLE
Localization / multi-language support?

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#8407] [Q&A]
Hi,

I'm trying to understand what multi-language support options there are with UNO.

Here is my starting point:
- all text resources are located in resources files: no hard coded text resources
- the shared project has a resource file Strings/en/ResourcesShared.**resw**
- there are additional .NET 6 libraries which have their resources in Resources.**resx** files located in the libraries

How would I need to structure both resources file types to enable an additional language (e.g. Spanish) so it would work on these target platforms?
- WindowsAppSdk
- UNO WASM
- UNO iOS
- UNO Android

Presumably I had to apply [the logic here](https://platform.uno/docs/articles/guides/localization.html?tabs=tabid-vswin), but I'm not sure how this would for **.resx** resource files.

Also: would I need to create seperate builds or is there a concept of 'satellite resource files' on the target platforms?

Thanks in advance

**Comment by jeromelaban** [comment]
You can take a look at [these samples](https://github.com/unoplatform/Uno.Samples/tree/master/UI/LocalizationSamples).

**Reply by DierkDroth** [reply]
Thanks @jeromelaban. I checked out the samples and confirmed what I already know, which is:
- how to structure **RESW** resource files in a shared project
- how to toggle languages at runtime / on startup

However, this stil won't tell me how to structure **RESX** resource files in .NET 6 libraries (**not** shared projects).

Anyway, I took the plunge and here is what I came up with (in case anyone else is facing that situation):
- **RESW** resources in shared are handled as demonstarted by @jeromelaban's samples
- however, there is problem with culture names: UNO (WASM) expects a culture identifier like 'en' as subdirectory name of the 'Strings' directory. 'en-US' (Windows standard) will give you trouble in case the app is running e.g. on a German machine: [no default resources would be found](https://github.com/unoplatform/uno/discussions/8351)
- if you're having additional .NET 6 libraries with **RESX** files (so not shared projects), then naming localized version of the resources as e.g. '.de.resx' - to follow the logic above - will fail for a WindowsAppSDK app on a German machine, since Windows expects '.de-DE.resx' in the resource file name (since the WindowsAppSDK culture name on a German machine is 'de-DE')

So, if you wanted to switch resources on startup or runtime then you have to be careful if you wanted to go with 'de' or 'de-DE'. For now, I'm going with 'de'. This means I have to map down any 'de-XX' to just 'de' (= new CultureInfo("de")) and anything which can't be mapped falls back to default (= new CultureInfo("en")).

**Reply by jeromelaban** [reply]
There are no directives on how to structure resx files in shared projects at this point, only resw files. This falls back onto the .NET [documentation on the topic](https://docs.microsoft.com/en-us/dotnet/core/extensions/work-with-resx-files-programmatically).

**Reply by DierkDroth** [reply]
@jeromelaban to clarify:
- with .NET 6 libraries (no UI involved, this is not a shared project) my only option is a **RESX** resource, since a **RESW** resource is not available for a .NET 6 target framework, correct?
- I just tested with 2 resources files 'Resources.resx' and 'Resources.de.resx' in a .NET 6 library and could get both to work as expected with WindowsAppSdk
- however, in UNO WASM it wasn't working and only the (english) default resources in 'Resources.resx' ever were pulled. I was not able to access the resources in 'Resources.de.resx'

This brings my back to my initial question: how could I get localized **RESX** files in .NET 6 libraries to work with UNO (WASM)?

**Reply by jeromelaban** [reply]
If this is .NET 6 libraries, Uno has nothing specific to do with it. If you have a repro to provide it'll help.

**Reply by DierkDroth** [reply]
Sure @jeromelaban. Please see MainPage.xaml.cs. 

Thanks

[Resources3.zip](https://github.com/unoplatform/uno/files/8334687/Resources3.zip)

**Reply by DierkDroth** [reply]
@jeromelaban have you been able to reproduce the issue? 

What could I do to get multi-language support in my UNO WASM app? Would you have any idea?

**Reply by DierkDroth** [reply]
Localization using resources is quite a nightmare with WASM (I suspect for the most part this falls on MS/Blazor's plate). 

Here is what I came up with (I have not checked iOS, Android at that point yet):
- as per my repo above RESX management isn't working at all. That's why I came up with a custom ResourcesManager which simply switches to a different RESX file based on CurrentUICulture - or just goes to the default RESX in case the CurrentUICulture is not supported
- this required that I implemented a custom tool to generate proper C# wrappers - basically replacing PublicResXFileCodeGenerator
- next I found that even RESW files in shared projects are not working properly since I saw cases where the default resource was not pulled if there was no localized version available
- this made the replacing the RESW (ResourceLoader) logic by the RESX (custom ResourceManager) logic from above, which provided me C# wrappers as well - perfect. BTW I never understood why MS came up with the RESW concept in the first place and did not just stick with RESX .... anyway ...
- I still had to work around cases where resources used on class property initializers did not work - they just had been blank. I suspect there is some race condition on app startup boiling...

I'm good to go for now and even would be able to deal with situations where I needed e.g. 2 different Spanisch CultureInfos (like es-ES and es-CL). I'll see how far my custom approach will take me...

**Reply by TonyK-FmGroup** [reply]
If I follow the thread correctly, are we saying that the localization strings dont work in WASM under .net 6?'
Under which version of net will it work? Is this expected to be fixed in .net any time soon?
I was about to show off a POC that I've been working on and it kinda fell over when I just tried the WASM and didn't get the localized content. Could actually be a show stopper for using Uno on a larger product.

**Reply by DierkDroth** [reply]
Well, it works. You just have to take a custom approach. After all resources in .NET are just XML files (no matter if RESX or RESW).

There is a chance I'm just missing something though. However, at some point I gave up digging and went with my said custom approach ... which works perfectly for me.

Note: I'm not doing any XAML at all - all my UI is in code behind (using an [approach similar to here](https://github.com/VincentH-Net/CSharpForMarkup)). Hence I did not try to get localized resources working in XAML.

**Reply by jeromelaban** [reply]
> If I follow the thread correctly, are we saying that the localization strings dont work in WASM under .net 6?' Under which version of net will it work? Is this expected to be fixed in .net any time soon? I was about to show off a POC that I've been working on and it kinda fell over when I just tried the WASM and didn't get the localized content. Could actually be a show stopper for using Uno on a larger product.

ResourceLoading is working properly with resw files. Resx is not supported at this time, as satellite assemblies are not published in the output package. Let's track this with https://github.com/unoplatform/uno/issues/8560.

The best way to localize for libraries for now is to provide an abstraction that will redirect to `Windows.ApplicationModel.Resources.ResourceLoader`.

**Reply by KWodarczyk** [reply]
@jeromelaban but in case of WASM how does that work ? it knows what language is set on client Os ? and also it it possible to change language in run time for wasm ?

**Reply by jeromelaban** [reply]
@KWodarczyk you can take a look at https://github.com/unoplatform/Uno.Samples/tree/master/UI/LocalizationSamples

**Reply by KWodarczyk** [reply]
Before I do look into it, does it change language based on client request? Because the scenario I'm talking about is when wasm is already hosted and requested from clients in different countries for example. Can this still work dynamical based on client country?

**Reply by jeromelaban** [reply]
Please take a look at the sample, it will answer all your questions.
